File: windows/ClipboardMaster.Core/HistoryPersistence.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ClipboardMaster.Core;

/// <summary>
/// 历史持久化：history.json 索引 + images/{uuid}.png 图片文件。
/// JSON 形状与 macOS 版完全一致（camelCase、kind 判别式、ISO8601 无小数秒），可跨平台互读。
/// 读取端逐条校验（损坏即丢弃该条）；损坏 JSON 备份后从空开始。
/// </summary>
public sealed class HistoryPersistence
{
    private const string IndexFileName = "history.json";
    private const string ImagesDirName = "images";
    private const int CurrentVersion = 1;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        Converters = { new KindDtoConverter(), new SwiftDateTimeConverter() },
    };

    private readonly string _directory;
    private readonly int _capacity;

    private string IndexPath => Path.Combine(_directory, IndexFileName);
    private string ImagesPath => Path.Combine(_directory, ImagesDirName);

    public HistoryPersistence(string directory, int capacity = HistoryLimits.Capacity)
    {
        _directory = directory;
        _capacity = Math.Max(1, capacity);
    }

    public IReadOnlyList<ClipboardEntry> Load()
    {
        if (!File.Exists(IndexPath)) return Array.Empty<ClipboardEntry>();
        PersistedIndex? index;
        try
        {
            index = JsonSerializer.Deserialize<PersistedIndex>(File.ReadAllText(IndexPath), JsonOptions);
        }
        catch (JsonException)
        {
            BackupCorruptIndex();
            return Array.Empty<ClipboardEntry>();
        }
        if (index?.Entries is null) return Array.Empty<ClipboardEntry>();
        return index.Entries
            .Select(ToEntry)
            .Where(e => e is not null)
            .Take(_capacity)
            .Cast<ClipboardEntry>()
            .ToList();
    }

    public void Save(IReadOnlyList<ClipboardEntry> entries)
    {
        Directory.CreateDirectory(ImagesPath);
        var dtos = entries.Select(ToDto).ToList();
        var index = new PersistedIndex(CurrentVersion, dtos);
        AtomicWrite(IndexPath, JsonSerializer.Serialize(index, JsonOptions));
        PruneOrphanImages(dtos);
    }

    public void ClearAll()
    {
        TryDelete(IndexPath);
        if (Directory.Exists(ImagesPath)) Directory.Delete(ImagesPath, true);
    }

    /// <summary>图片条目对应的磁盘文件路径（与保存命名一致：{uuid}.png）；非图片或未落盘返回 null。</summary>
    public string? GetImageFilePath(ClipboardEntry entry)
    {
        if (entry.Content is not ImageContent) return null;
        var path = Path.Combine(ImagesPath, $"{entry.Id.ToString("D")}.png");
        return File.Exists(path) ? path : null;
    }

    // ---- 内部转换与校验 ----

    private ClipboardEntry? ToEntry(PersistedEntry dto) => dto.Kind switch
    {
        TextKindDto t => ClipboardContent.Make(t.Text) is { } c
            ? new ClipboardEntry(dto.Id, dto.CapturedAt, c)
            : null,
        ImageKindDto img => LoadImage(img) is { } c2
            ? new ClipboardEntry(dto.Id, dto.CapturedAt, c2)
            : null,
        _ => null,
    };

    private ImageContent? LoadImage(ImageKindDto dto)
    {
        var path = Path.Combine(ImagesPath, dto.File);
        if (!File.Exists(path)) return null;
        try
        {
            var bytes = File.ReadAllBytes(path);
            return ClipboardContent.Make(bytes, dto.Width, dto.Height);
        }
        catch (IOException)
        {
            return null;
        }
    }

    private PersistedEntry ToDto(ClipboardEntry entry) => entry.Content switch
    {
        TextContent t => new PersistedEntry(entry.Id, entry.CapturedAt, new TextKindDto(t.Text)),
        ImageContent img => new PersistedEntry(entry.Id, entry.CapturedAt, SaveImage(entry.Id, img)),
        _ => throw new InvalidOperationException("未知内容类型"),
    };

    private ImageKindDto SaveImage(Guid id, ImageContent img)
    {
        // 文件名与条目 Id 绑定：重复保存同一条目不产生新文件（与 mac 版一致）。
        var fileName = $"{id.ToString("D")}.png";
        AtomicWrite(Path.Combine(ImagesPath, fileName), img.ImageData);
        return new ImageKindDto(fileName, img.PixelWidth, img.PixelHeight);
    }

    private void PruneOrphanImages(List<PersistedEntry> dtos)
    {
        var referenced = dtos
            .Select(d => d.Kind as ImageKindDto)
            .Where(k => k is not null)
            .Select(k => k!.File)
            .ToHashSet();
        if (!Directory.Exists(ImagesPath)) return;
        foreach (var file in Directory.GetFiles(ImagesPath))
        {
            if (!referenced.Contains(Path.GetFileName(file))) TryDelete(file);
        }
    }

    private void BackupCorruptIndex()
    {
        var backupPath = IndexPath + ".corrupt";
        TryDelete(backupPath);
        try
        {
            File.Move(IndexPath, backupPath);
        }
        catch (IOException)
        {
            // 备份失败则保留原文件，下次启动重试
        }
    }

    private static void AtomicWrite(string path, string text) => AtomicWrite(path, Encoding.UTF8.GetBytes(text));

    private static void AtomicWrite(string path, byte[] bytes)
    {
        var tmpPath = path + ".tmp";
        File.WriteAllBytes(tmpPath, bytes);
        File.Move(tmpPath, path, overwrite: true);
    }

    private static void TryDelete(string path)
    {
        try { File.Delete(path); }
        catch (IOException) { }
    }
}

// ---- 持久化 DTO（形状与 Swift 版 Codable 输出逐字节对齐）----

internal sealed record PersistedIndex(int Version, List<PersistedEntry> Entries);

internal sealed record PersistedEntry(Guid Id, DateTime CapturedAt, KindDto Kind);

internal abstract record KindDto;

internal sealed record TextKindDto(string Text) : KindDto;

internal sealed record ImageKindDto(string File, int Width, int Height) : KindDto;

/// <summary>判别式联合：{"text":{"text":...}} / {"image":{"file":...,"width":...,"height":...}}</summary>
internal sealed class KindDtoConverter : JsonConverter<KindDto>
{
    public override KindDto Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        using var doc = JsonDocument.ParseValue(ref reader);
        var root = doc.RootElement;
        if (root.ValueKind != JsonValueKind.Object || root.EnumerateObject().Count() != 1)
            throw new JsonException("kind 应为单属性判别对象");
        var prop = root.EnumerateObject().Single();
        var value = prop.Value;
        return prop.Name switch
        {
            "text" => new TextKindDto(value.GetProperty("text").GetString() ?? ""),
            "image" => new ImageKindDto(
                value.GetProperty("file").GetString() ?? "",
                value.GetProperty("width").GetInt32(),
                value.GetProperty("height").GetInt32()),
            _ => throw new JsonException($"未知 kind: {prop.Name}"),
        };
    }

    public override void Write(Utf8JsonWriter writer, KindDto value, JsonSerializerOptions options)
    {
        writer.WriteStartObject();
        switch (value)
        {
            case TextKindDto t:
                writer.WritePropertyName("text");
                writer.WriteStartObject();
                writer.WriteString("text", t.Text);
                writer.WriteEndObject();
                break;
            case ImageKindDto img:
                writer.WritePropertyName("image");
                writer.WriteStartObject();
                writer.WriteString("file", img.File);
                writer.WriteNumber("width", img.Width);
                writer.WriteNumber("height", img.Height);
                writer.WriteEndObject();
                break;
            default:
                throw new JsonException("未知 KindDto");
        }
        writer.WriteEndObject();
    }
}

/// <summary>Swift .iso8601 兼容：写入无小数秒的 UTC；读取兼容任意 ISO8601。</summary>
internal sealed class SwiftDateTimeConverter : JsonConverter<DateTime>
{
    private const string Format = "yyyy-MM-dd'T'HH:mm:ss'Z'";

    public override DateTime Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        var s = reader.GetString()
            ?? throw new JsonException("日期应为字符串");
        return DateTime.Parse(s, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind).ToUniversalTime();
    }

    public override void Write(Utf8JsonWriter writer, DateTime value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.ToUniversalTime().ToString(Format, CultureInfo.InvariantCulture));
    }
}

File: Sources/ClipboardMasterCore/HistoryPersistence.swift
import Foundation

/// 历史持久化：`history.json` 索引 + `images/<uuid>.png` 图片文件。
/// 读取端逐条校验（损坏即丢弃该条，不影响其余）；损坏 JSON 备份后从空开始。
public final class HistoryPersistence {
    private let fileManager = FileManager.default
    private let directory: URL
    private let capacity: Int

    private var indexURL: URL { directory.appendingPathComponent("history.json") }
    private var imagesURL: URL { directory.appendingPathComponent("images") }

    public init(directory: URL, capacity: Int = HistoryLimits.capacity) {
        self.directory = directory
        self.capacity = max(1, capacity)
    }

    // MARK: - 读取

    public func load() -> [ClipboardEntry] {
        guard let data = try? Data(contentsOf: indexURL) else { return [] }
        guard let persisted = decodeIndex(data) else {
            backupCorruptIndex()
            return []
        }
        let entries = persisted.entries.compactMap(entryFromDTO)
        return Array(entries.prefix(capacity))
    }

    // MARK: - 写入

    public func save(_ entries: [ClipboardEntry]) throws {
        try fileManager.createDirectory(at: imagesURL, withIntermediateDirectories: true)
        let dtos = try entries.map(dtoFromEntry)
        let index = PersistedIndex(version: PersistedIndex.currentVersion, entries: dtos)
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.sortedKeys]
        let data = try encoder.encode(index)
        try data.write(to: indexURL, options: .atomic)
        pruneOrphanImages(referencedBy: dtos)
    }

    public func clearAll() throws {
        try? fileManager.removeItem(at: indexURL)
        try? fileManager.removeItem(at: imagesURL)
    }

    /// 图片条目对应的磁盘文件路径（与保存命名一致：<uuid>.png）；非图片或未落盘返回 nil。
    public func imageFileURL(for entry: ClipboardEntry) -> URL? {
        guard case .image = entry.content else { return nil }
        let url = imagesURL.appendingPathComponent("\(entry.id.uuidString).png")
        return fileManager.fileExists(atPath: url.path) ? url : nil
    }

    /// 目录级迁移（v1 ClipHistory → v2 ClipboardMaster）：
    /// 仅当源存在且目标不存在时整体移动；其余情况保持不动（幂等、不覆盖用户新数据）。
    public static func migrateLegacyDirectory(from source: URL, to destination: URL) {
        let fm = FileManager.default
        guard fm.fileExists(atPath: source.path),
              !fm.fileExists(atPath: destination.path)
        else { return }
        do {
            try fm.createDirectory(at: destination.deletingLastPathComponent(), withIntermediateDirectories: true)
            try fm.moveItem(at: source, to: destination)
        } catch {
            FileHandle.standardError.write(
                Data("[ClipboardMaster] 历史目录迁移失败（保留原目录）: \(error)\n".utf8)
            )
        }
    }

    // MARK: - 私有：编解码

    private func decodeIndex(_ data: Data) -> PersistedIndex? {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try? decoder.decode(PersistedIndex.self, from: data)
    }

    private func backupCorruptIndex() {
        let backupURL = directory.appendingPathComponent("history.json.corrupt")
        try? fileManager.removeItem(at: backupURL)
        try? fileManager.moveItem(at: indexURL, to: backupURL)
        FileHandle.standardError.write(
            Data("[ClipHistory] history.json 无法解码，已备份为 history.json.corrupt，历史从空开始\n".utf8)
        )
    }

    private func entryFromDTO(_ dto: PersistedEntry) -> ClipboardEntry? {
        switch dto.kind {
        case .text(let text):
            guard let content = ClipboardContent.make(text: text) else { return nil }
            return ClipboardEntry(id: dto.id, capturedAt: dto.capturedAt, content: content)
        case .image(let file, let width, let height):
            let url = imagesURL.appendingPathComponent(file)
            guard let imageData = try? Data(contentsOf: url),
                  let content = ClipboardContent.make(imageData: imageData, pixelWidth: width, pixelHeight: height)
            else { return nil }
            return ClipboardEntry(id: dto.id, capturedAt: dto.capturedAt, content: content)
        }
    }

    private func dtoFromEntry(_ entry: ClipboardEntry) throws -> PersistedEntry {
        switch entry.content {
        case .text(let text):
            return PersistedEntry(id: entry.id, capturedAt: entry.capturedAt, kind: .text(text: text))
        case .image(let imageData, let width, let height):
            // 文件名与条目 id 绑定：重复保存同一 id 不产生新文件。
            let fileName = "\(entry.id.uuidString).png"
            let fileURL = imagesURL.appendingPathComponent(fileName)
            try imageData.write(to: fileURL, options: .atomic)
            return PersistedEntry(id: entry.id, capturedAt: entry.capturedAt, kind: .image(file: fileName, width: width, height: height))
        }
    }

    private func pruneOrphanImages(referencedBy dtos: [PersistedEntry]) {
        let referenced = Set(dtos.compactMap { dto -> String? in
            if case .image(let file, _, _) = dto.kind { return file }
            return nil
        })
        guard let files = try? fileManager.contentsOfDirectory(at: imagesURL, includingPropertiesForKeys: nil) else {
            return
        }
        for file in files where !referenced.contains(file.lastPathComponent) {
            try? fileManager.removeItem(at: file)
        }
    }
}

// MARK: - 持久化 DTO（显式 Codable，与领域模型解耦，便于逐字段校验）

private struct PersistedIndex: Codable {
    static let currentVersion = 1

    let version: Int
    let entries: [PersistedEntry]
}

private struct PersistedEntry: Codable {
    enum Kind: Codable {
        case text(text: String)
        case image(file: String, width: Int, height: Int)
    }

    let id: UUID
    let capturedAt: Date
    let kind: Kind
}
